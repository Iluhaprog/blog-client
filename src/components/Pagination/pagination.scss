@import '../../vars.scss';

.pagination {
    font-size: 14px;
    .button {
        min-width: 30px;
        width: 30px;
    }

    a {
        text-decoration: none;
        font-size: 20px;
        padding: 5px 15px;
        margin: 5px;
        color: $text-color-6;
        border: $border-2;
        border-radius: 5px;
        transition: .3s ease-in-out;

        &.selected-page {
            color: $nav-bg;
            font-weight: 600;
            border: $border-1;
        }
    }
}