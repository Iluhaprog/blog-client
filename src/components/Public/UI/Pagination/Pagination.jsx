import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import styles from './Pagination.module.css';

function Pagination({total, limit, onClick}) {
  const pageCount = Math.ceil(total / limit);
  return (
    <ReactPaginate
      previousLabel={'<'}
      nextLabel={'>'}
      breakLabel={'...'}
      containerClassName={styles.pagination}
      pageClassName={styles.pageItem}
      previousClassName={styles.pageItem}
      nextClassName={styles.pageItem}
      breakClassName={styles.pageItem}
      breakLinkClassName={styles.pageLink}
      previousLinkClassName={styles.pageLink}
      nextLinkClassName={styles.pageLink}
      pageLinkClassName={styles.pageLink}
      activeClassName={styles.active}
      pageCount={pageCount}
      pageRangeDisplay={3}
      marginPagesDisplayed={1}
      onPageChange={onClick}
    />
  );
}

Pagination.propTypes = {
  total: PropTypes.number,
  limit: PropTypes.number,
  onClick: PropTypes.func,
};

export {Pagination};
