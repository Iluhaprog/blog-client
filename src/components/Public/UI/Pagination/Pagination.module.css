.pagination {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    list-style: none;
    max-width: fit-content;
    margin: 0 auto;
    padding: 0;
}

.pageLink {
    font-family: var(--public-font);
    border: 1px solid var(--color-1);
    width: 35px;
    height: 35px;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 50%;
    margin: 5px;
}

.pageItem:first-child {
    margin-right: 15px;
}

.pageItem:last-child {
    margin-left: 15px;
}

.pageLink {
    color: var(--color-1);
    outline: none;
}

.pageLink:hover,
.pageItem.active .pageLink {
    color: #fff;
    background-color: var(--color-1);
    text-decoration: none;
}
