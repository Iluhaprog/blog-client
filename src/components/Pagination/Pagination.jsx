import React, { useEffect, useState } from "react"
import { NavLink } from "react-router-dom";
import { PrimaryButton } from "../buttons";
import { Row } from '../containers';
import './pagination.scss';

function Pagination({ totalItems, itemsPerPage, currentPage, changePage, visiblePageLinks = 3, page }) {
    const visiblePageLinksIndexInit = Math.floor((currentPage - 1) / visiblePageLinks);
    const [visiblePageLinksIndex, setVisiblePageLinksIndex] = useState(visiblePageLinksIndexInit);
    useEffect(() => {
        setVisiblePageLinksIndex(visiblePageLinksIndexInit);
    }, [currentPage])
    const totalPagesLength = Math.ceil(totalItems / itemsPerPage);
    const visiblePageLinksIndexTotal = Math.ceil(totalPagesLength / visiblePageLinks);

    const totalPages = totalPagesLength < visiblePageLinks 
            ? totalPagesLength 
            : visiblePageLinksIndex + 1 === visiblePageLinksIndexTotal 
                ? visiblePageLinks - Math.abs(visiblePageLinks * visiblePageLinksIndexTotal - totalPagesLength)
                : visiblePageLinks;
    const pages = Array(totalPages).fill(0);
    return (
        <div className='pagination'>
            <Row justifyContent='c' alignItems='c'>
                {
                    visiblePageLinksIndex 
                        ? (
                            <PrimaryButton 
                                onClick={() => setVisiblePageLinksIndex(visiblePageLinksIndex - 1)}
                                text='<'/>
                        ) 
                        : ''
                }
                {
                    pages.map((_, index) => (
                        <NavLink 
                            key={index} 
                            to={`/admin/${page}/${index + 1 + visiblePageLinksIndex * visiblePageLinks}`}
                            onClick={() => changePage(index + visiblePageLinksIndex * visiblePageLinks)}
                            className={
                                index + 1 + visiblePageLinksIndex * visiblePageLinks === currentPage 
                                    ? 'selected-page'
                                    : '' 
                            }
                        >
                            {index + 1 + visiblePageLinksIndex * visiblePageLinks}
                        </NavLink>
                    ))
                }
                {
                    (totalPagesLength > visiblePageLinks && (visiblePageLinksIndex + 1) * visiblePageLinks < totalPagesLength ) && (
                        <PrimaryButton 
                            onClick={() => setVisiblePageLinksIndex(visiblePageLinksIndex + 1)}
                            text='>'
                        />
                    )
                }
            </Row>
        </div>
    )
}

export default Pagination;