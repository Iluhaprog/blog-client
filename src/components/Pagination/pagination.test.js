import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Pagination from './Pagination';

test('test view of pagination first', () => {
    const { asFragment } = render(
        <BrowserRouter>
            <Pagination
                totalItems={100}
                itemsPerPage={10}
                currentPage={1}
                changePage={() => {}}
                visiblePageLinks={5}
                page='test'
            />
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test('test view of pagination center', () => {
    const { asFragment } = render(
        <BrowserRouter>
            <Pagination
                totalItems={100}
                itemsPerPage={10}
                currentPage={5}
                changePage={() => {}}
                visiblePageLinks={3}
                page='test'
            />
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test('test view of pagination last page', () => {
    const { asFragment } = render(
        <BrowserRouter>
            <Pagination
                totalItems={100}
                itemsPerPage={10}
                currentPage={10}
                changePage={() => {}}
                visiblePageLinks={3}
                page='test'
            />
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test('test change page of pagination', async () => {
    const handleChangePage = jest.fn();
    const { getByText } = render(
        <BrowserRouter>
            <Pagination
                totalItems={100}
                itemsPerPage={10}
                currentPage={1}
                changePage={handleChangePage}
                visiblePageLinks={5}
                page='test'
            />
        </BrowserRouter>
    );
    fireEvent.click(getByText('1'));
    await waitFor(() => {
        expect(handleChangePage).toHaveBeenCalledTimes(1);
    })
});