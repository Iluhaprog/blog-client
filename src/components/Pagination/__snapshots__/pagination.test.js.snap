// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test view of pagination center 1`] = `
<DocumentFragment>
  <div
    class="pagination"
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <button
        class="button default-button primary-button"
      >
        &lt;
      </button>
      <a
        class=""
        href="/admin/posts/4"
      >
        4
      </a>
      <a
        class="selected-page"
        href="/admin/posts/5"
      >
        5
      </a>
      <a
        class=""
        href="/admin/posts/6"
      >
        6
      </a>
      <button
        class="button default-button primary-button"
      >
        &gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test view of pagination first 1`] = `
<DocumentFragment>
  <div
    class="pagination"
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <a
        class="selected-page"
        href="/admin/posts/1"
      >
        1
      </a>
      <a
        class=""
        href="/admin/posts/2"
      >
        2
      </a>
      <a
        class=""
        href="/admin/posts/3"
      >
        3
      </a>
      <a
        class=""
        href="/admin/posts/4"
      >
        4
      </a>
      <a
        class=""
        href="/admin/posts/5"
      >
        5
      </a>
      <button
        class="button default-button primary-button"
      >
        &gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test view of pagination last page 1`] = `
<DocumentFragment>
  <div
    class="pagination"
  >
    <div
      class="row row_jc-c row_ai-c "
    >
      <button
        class="button default-button primary-button"
      >
        &lt;
      </button>
      <a
        class="selected-page"
        href="/admin/posts/10"
      >
        10
      </a>
    </div>
  </div>
</DocumentFragment>
`;
